// bunker/index.js
var burrito = require('burrito');
var vm = require('vm');
var EventEmitter = require('events').EventEmitter;

module.exports = function (src) {
    var b = new Bunker();
    if (src) b.include(src);
    return b;
};

function Bunker () {
    this.src = {};
    this.mut = {};
    this.nodes = [];
    
    this.names = {
        call : burrito.generateName(6),
        expr : burrito.generateName(6),
        stat : burrito.generateName(6)
    };
}

Bunker.prototype = new EventEmitter;

Bunker.prototype.addTest = function (test) {
    this.test = test;
    return this;
};

Bunker.prototype.addSource = function (src,mut) {
    this.src = src;
    this.source = null;
    this.mut = mut;
    return this;
};

Bunker.prototype.compile = function () {
    var src = this.src;
    var mut = this.mut;
    var nodes = this.nodes;
    var names = this.names;
    return burrito(src, function (node,mut) {
        var i = nodes.length;
        console.log("NODE");
        console.log(node.name) + " " + console.log(node.label()) + " " + console.log(node.source());
        if (node.name === 'call') {
            nodes.push(node);
            node.wrap(names.call + '(' + i + ')(%s)');
        }
        else if (node.name === 'stat' || node.name === 'throw'
        || node.name === 'var') {
            nodes.push(node);
            node.wrap('{' + names.stat + '(' + i + ');%s}');
        }
        else if (node.name === 'binary') {
            nodes.push(node);
            node.wrap(names.expr + '(' + i + ')(%s)');
        }
        else if (node.name === 'unary-postfix' || node.name === 'unary-prefix') {
            nodes.push(node);
            node.wrap(names.expr + '(' + i + ')(%s)');
        }
        if (i !== nodes.length) {
            node.id = i;
        }
    }) + this.test;
};

Bunker.prototype.assign = function (context) {
    if (!context) context = {};
    
    var self = this;
    var stack = [];
    
    context[self.names.call] = function (i) {
        var node = self.nodes[i];
        stack.unshift(node);
        self.emit('node', node, stack);
        
        return function (expr) {
            stack.shift();
            return expr;
        };
    };
    
    context[self.names.expr] = function (i) {
        var node = self.nodes[i];
        self.emit('node', node, stack);
        
        return function (expr) {
            return expr;
        };
    };
    
    context[self.names.stat] = function (i) {
        var node = self.nodes[i];
        self.emit('node', node, stack);
    };
    
    return context;
};
    
Bunker.prototype.run = function (context) {
    var src = this.compile();
//    vm.runInNewContext(src, this.assign(context));
    vm.runInNewContext(src, this.assign(context));    
    return this;
};
